### ISSUE TITLE
Add support for ReadSeeker for uploads to s3

### ISSUE THREAD

**Issue by mskonovalov** [vbauerster/mpb#114]
Unfortunately s3 sdk uses ReadSeeker for uploads.
So in this case it is tricky to wrap a file that you trying to upload: curremt API only allows ReadCloser but not ReadSeeker.

**Comment by mskonovalov** [comment]
@vbauerster I suspect they use ReadSeeker to be able to re-upload specific parts of the file.
That's exactly what I do atm:

```
ReadSeeker{
	file,
	p.ProxyReader(file),
}
```
but I suspect this will show wrong results in case you try to re-upload the specific part of the file, it easily can get higher 100%

**Comment by vbauerster** [comment]
> but I suspect this will show wrong results in case you try to re-upload the specific part of the file, it easily can get higher 100%

That's correct. There is [SetCurrent](https://github.com/vbauerster/mpb/blob/7e6c4e404ee7b01e3f8d095b0263c7ed5956d43b/bar.go#L210) which may help to keep bar in correct state after seek has been called. If you come up with working implementation, PR is welcome. I'm not able to test against s3 api.

**Comment by mskonovalov** [comment]
So I started to test and unfortunately it doesn't work completely:
AWS SDK reads body multiple times to calculate different hashes and so on :(
Because the reader is seekable they can do it without issues.
Don't have ideas yet how to solve it

**Comment by mskonovalov** [comment]
Ok, I was able to come up with some weird solution:
```

type ProxyReadSeeker struct {
	F           io.ReadSeekCloser
	ProxyReader io.ReadCloser
	Bar         *mpb.Bar
}

func (rs *ProxyReadSeeker) Read(p []byte) (n int, err error) {
	return rs.ProxyReader.Read(p)
}

func (rs *ProxyReadSeeker) Close() error {
	return rs.F.Close()
}

func (rs *ProxyReadSeeker) Seek(offset int64, whence int) (int64, error) {
	if whence == io.SeekStart {
		rs.Bar.SetCurrent(offset)
	}
	return rs.F.Seek(offset, whence)
}
```
Not sure if it is useful for anyone else, but it seems to work. Although you cannot rely on automatic completion and have to disable it and complete the bar manually.

**Comment by Solaris-star** [comment]
This is a legitimate gap — the AWS SDK v2 `Upload` API requires `io.ReadSeeker` for retry support, and `ProxyReader` only wraps `io.Reader`.

A `ProxyReadSeeker` that wraps `io.ReadSeeker` and forwards `Seek` calls (resetting the byte counter on seek-to-start) would cover this. The implementation is straightforward: embed the existing ProxyReader logic and add a `Seek` method that calls through to the underlying seeker.

I can take a crack at this if there's interest from @vbauerster.

**Comment by vbauerster** [comment]
Yes please, but I won't be able to test it against real AWS SDK so do you crack here.😉
